### PR TITLE
[processing] support for scripts in the test suite

### PR DESCRIPTION
**Pull request by alexbruy**
Adds ability to run tests against Processing scripts to make sure that scripts provider works correctly.

Scripts should be created and saved in the `scripts` subdirectory of the Processing `testdata` directory. Script file name should be the same as name of the script algorithm in the Processing toolbox. To create test follow the same rules as for ordinal Processing algorithms.

Small test script (centroids calculation) included.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

File: python/plugins/processing/tests/testdata/scripts/centroids.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+for count, f in enumerate(features):
+    outputFeature = f
+    if f.geometry():
+        outputGeometry = f.geometry().centroid()
+        outputFeature.setGeometry(outputGeometry)
+
+    writer.addFeature(outputFeature)
+    feedback.setProgress(int(count * total))
+
+del writer
```
Comment (m-kuhn) [inline python/plugins/processing/tests/testdata/scripts/centroids.py]: Shouldn't that be garbage collected anyway?

File: python/plugins/processing/tests/testdata/scripts/centroids.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+features = processing.features(layer)
+count = len(features)
+if count == 0:
+    raise GeoAlgorithmExecutionException('Input layer contains no features.')
+
+total = 100.0 / len(features)
+
+for count, f in enumerate(features):
+    outputFeature = f
+    if f.geometry():
```
Comment (m-kuhn) [inline python/plugins/processing/tests/testdata/scripts/centroids.py]: `f.hasGeometry()` should be a tiny bit faster

**Comment by m-kuhn** [comment]
Looks good to me.
I don't know scripts very well. They are mostly user-written (i.e. not shipped) so it's really about testing the provider with scripts using some of its different features rather than the scripts themselves?

**Comment by alexbruy** [comment]
Right, scripts mostly user-written. This PR has two goals:

- test scripts provider, so we are sure that it works and scripts loaded and executed correctly
- help with testing algorithms which produce no output (like selecting features) by combining them with another algorithm which can create output. So for example, we can create script which will call "select by attribute" algorithm with some layer, and then use "save selected features" algorithm to produce output. This output then can be compared with expected data.

Hope this makes sense.

File: python/plugins/processing/tests/testdata/scripts/centroids.py (hunk 1 above)
Comment (alexbruy) [inline python/plugins/processing/tests/testdata/scripts/centroids.py]: You are right

**Comment by m-kuhn** [comment]
Makes a lot of sense to me, thanks.

Could you add (or link to) a script that calls another algorithm that can serve as an exmaple?

Where would you put those "wrapper tests"? For people looking for tests in the future I think they would be be put directly into the qgis / gdal etc. yamls?

**Review by m-kuhn — APPROVED** [review]

**Comment by m-kuhn** [comment]
Good move!

**Comment by alexbruy** [comment]
@m-kuhn I added sample script which tests "Select By Attribute" algorithm using approach I described above. Such scripts can be created manually or using Modeler's "export to Python" functionality.

Regarding placement I think if script used to tests QGIS or GDAL algorithm it should go to the corresponding `.yaml` file. Dedicated scripts `.yaml` should be used for scripts which test scripts provider itself. If we don't need separate tests, we can remove it and use existing files.

**Pull request merged**